common/models/checkpoint.js

```diff
@@ -27,43 +28,59 @@ module.exports = function(Checkpoint) {
    * Get the current checkpoint id
    * @callback {Function} callback
    * @param {Error} err
-   * @param {Number} checkpointId The current checkpoint id
+   * @param {Number} checkpoint The current checkpoint seq
    */
-
   Checkpoint.current = function(cb) {
     var Checkpoint = this;
-    this.find({
-      limit: 1,
-      order: 'seq DESC'
-    }, function(err, checkpoints) {
-      if (err) return cb(err);
-      var checkpoint = checkpoints[0];
-      if (checkpoint) {
-        cb(null, checkpoint.seq);
-      } else {
-        Checkpoint.create({ seq: 1 }, function(err, checkpoint) {
-          if (err) return cb(err);
-          cb(null, checkpoint.seq);
-        });
+
+    async.waterfall([
+      function(cb) {
+        Checkpoint.findOne(cb);
+      },
+      function(cp, cb) {
+        if (cp) {
+          cb(null, cp.seq);
+        } else {
+          Checkpoint.init(function(err, data) {
+            cb(err, data.seq);
+          });
+        }
       }
-    });
+    ],
+    cb);
   };
 
-  Checkpoint.observe('before save', function(ctx, next) {
-    if (!ctx.instance) {
-      // Example: Checkpoint.updateAll() and Checkpoint.updateOrCreate()
-      return next(new Error('Checkpoint does not support partial updates.'));
-    }
+  /**
+   * Initialize the checkpoint
+   * @callback {Function} callback
+   * @param {Error} err
+   * @param {Object} checkpoint The current checkpoint
+   */
+  Checkpoint.init = function(cb) {
+    this.create({seq: 1}, cb);
+  };
 
-    var model = ctx.instance;
-    if (!model.getId() && model.seq === undefined) {
-      model.constructor.current(function(err, seq) {
-        if (err) return next(err);
-        model.seq = seq + 1;
-        next();
-      });
-    } else {
-      next();
-    }
-  });
+  /**
+   * Increase the current checkpoint If it already exists otherwise initialize it
+   * @callback {Function} callback
+   * @param {Error} err
+   * @param {Object} checkpoint The current checkpoint
+   */
+  Checkpoint.bump = function(cb) {
+    var Checkpoint = this;
+
+    async.waterfall([
+      function(cb) {
+        Checkpoint.findOne(cb);
+      },
+      function(cp, cb) {
+        if (!cp) {
+          Checkpoint.init(cb);
+        } else {
+          cp.seq++;
+          cp.save(cb);
+        }
+      }
+    ], cb);
```

Comment:
Right; this semantic  (`updateOrCreate(filter, { seq: { $inc: 1 } }, { seq: 1})`)  seems the only one which solves the problem of race condition in our case...